main/src/ca/uwaterloo/flix/api/Bootstrap.scala

```diff
@@ -336,106 +334,60 @@ class Bootstrap(val projectPath: Path, apiKey: Option[String]) {
-      val resourcesDir = Bootstrap.getResourcesDirectory(projectPath)
-      val resources = FileOps.getFilesIn(resourcesDir, Int.MaxValue)
-      for ((resource, fileNameWithSlashes) <- FileOps.sortPlatformIndependently(resourcesDir, resources)) {
-        FileOps.addToZip(zip, fileNameWithSlashes, resource)
-      }
-
-      if (includeDependencies) {
-        // First, we get all jar files inside the lib folder.
-        // If the lib folder doesn't exist, we suppose there is simply no dependency and trigger no error.
-        val jarDependencies = if (libDir.toFile.exists()) FileOps.getFilesWithExtIn(libDir, EXT_JAR, Int.MaxValue) else List[Path]()
-        // Add jar dependencies.
-        jarDependencies.foreach(dep => {
-          if (!Bootstrap.isJarFile(dep))
-            return Validation.Failure(BootstrapError.FileError(s"The jar file '${dep.toFile.getName} seems corrupted. Refusing to build fatjar-file."))
-
-          // Extract the content of the classes to the jar file.
-          Using(new ZipInputStream(Files.newInputStream(dep))) {
-            zipIn =>
-              var entry = zipIn.getNextEntry
-              while (entry != null) {
-                // Get the class files except module-info and META-INF classes which are specific to each library.
-                if (entry.getName.endsWith(s".$EXT_CLASS") && !entry.getName.equals(s"module-info.$EXT_CLASS") && !entry.getName.contains("META-INF/")) {
-                  // Write extracted class files to zip.
-                  val classContent = zipIn.readAllBytes()
-                  FileOps.addToZip(zip, entry.getName, classContent)
-                }
-                entry = zipIn.getNextEntry
-              }
-          }
-        })
-      }
-    } match {
-      case Success(()) => Validation.Success(())
-      case Failure(e) => Validation.Failure(BootstrapError.GeneralError(List(e.getMessage)))
+    Steps.compile(flix) match {
+      case Ok(result) => Validation.Success(result)
+      case Err(e) => Validation.Failure(e)
     }
   }
 
   /**
     * Builds a jar package for the project.
     */
   def buildJar(flix: Flix)(implicit formatter: Formatter): Validation[Unit, BootstrapError] = {
-    buildJarBase(flix, includeDependencies = false)
+    val jarFile = Bootstrap.getJarFile(projectPath)
+    Steps.updateStaleSources(flix)
+    val result = for {
+      _ <- Steps.configureJarOutput(flix)
+      _ <- Steps.compile(flix)
+      _ <- Steps.validateJarFile(jarFile)
+      _ = Files.createDirectories(getArtifactDirectory(projectPath))
+      contents = sequence(Seq(
+        Steps.addManifestToZip,
+        Steps.addClassFilesFromDirToZip(Bootstrap.getClassDirectory(projectPath)),
+        Steps.addResourcesFromDirToZip(Bootstrap.getResourcesDirectory(projectPath))
+      ))
+      _ <- toResult(Using(new ZipOutputStream(Files.newOutputStream(jarFile)))(contents))
+    } yield {
+      ()
+    }
+    result.toValidation
```

Comment:
Why the local var?

Reply:
To convert to validation. You can't invoke the method directly on the for-comprehension

Reply:
I see. Not even after the brace? You would need parens?

Reply:
Yes exactly.